### PR TITLE
Fixes Apple Calendar onboarding and type fixes

### PR DESCRIPTION
**Pull request by zomars**
- Fixed adding Apple calendars from onboarding, and took the liberty to...
- Refactor the whole onboarding calendar connecting experience to match the new integrations facelift

### CODE REVIEW

File: components/integrations/CalendarSwitch.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
+import { useMutation } from "react-query";
```
Comment (zomars) [inline components/integrations/CalendarSwitch.tsx:1]: Needed to extract all these components to reuse on onboarding @KATT

File: components/ui/UsernameInput.tsx [hunk 2]
```diff
@@ -1,6 +1,10 @@
 import React from "react";
 
-const UsernameInput = React.forwardRef((props, ref) => (
+interface UsernameInputProps extends React.ComponentPropsWithRef<"input"> {
+  label?: string;
+}
+
+const UsernameInput = React.forwardRef<HTMLInputElement, UsernameInputProps>((props, ref) => (
```
Comment (zomars) [inline components/ui/UsernameInput.tsx:7]: Type fixes

File: components/ui/WeekdaySelect.tsx [hunk 3]
```diff
@@ -1,8 +1,13 @@
...
-export const WeekdaySelect = (props) => {
+interface WeekdaySelectProps {
+  defaultValue: number[];
+  onSelect: (selected: number[]) => void;
+}
+
+export const WeekdaySelect = (props: WeekdaySelectProps) => {
   const [activeDays, setActiveDays] = useState(
-    [...Array(7).keys()].map((v, i) => (props.defaultValue || []).includes(i))
+    Array.from(Array(7).keys()).map((v, i) => (props.defaultValue || []).includes(i))
```
Comment (zomars) [inline components/ui/WeekdaySelect.tsx:10]: Type fixes

File: components/ui/form/radio-area/RadioAreaGroup.tsx [hunk 4]
```diff
@@ -1,10 +1,10 @@
...
+import React, { ReactNode, useState } from "react";
 
 import classNames from "@lib/classNames";
 
-type RadioAreaProps = React.InputHTMLAttributes<HTMLInputElement> & {
-  onChange: (value: string) => void;
-  defaultChecked: boolean;
+type RadioAreaProps = Omit<React.InputHTMLAttributes<HTMLInputElement>, "onChange"> & {
+  onChange?: (value: string) => void;
+  defaultChecked?: boolean;
```
Comment (zomars) [inline components/ui/form/radio-area/RadioAreaGroup.tsx:7]: This whole file was a tricky one to fix types

File: ee/pages/api/integrations/stripepayment/webhook.ts [hunk 5]
```diff
@@ -1,12 +1,12 @@
 import { buffer } from "micro";
 import type { NextApiRequest, NextApiResponse } from "next";
-import { getErrorFromUnknown } from "pages/_error";
 import Stripe from "stripe";
 
 import stripe from "@ee/lib/stripe/server";
 
 import { CalendarEvent } from "@lib/calendarClient";
 import { HttpError } from "@lib/core/http/error";
+import { getErrorFromUnknown } from "@lib/errors";
```
Comment (zomars) [inline ee/pages/api/integrations/stripepayment/webhook.ts:9]: Extracted to avoid importing from `pages`

File: lib/core/i18n/i18n.utils.ts [hunk 6]
```diff
@@ -9,8 +9,10 @@ import { Maybe } from "@trpc/server";
 import { i18n } from "../../../next-i18next.config";
 
 export function getLocaleFromHeaders(req: IncomingMessage): string {
-  const preferredLocale = parser.pick(i18n.locales, req.headers["accept-language"]) as Maybe<string>;
-
+  let preferredLocale: string | null | undefined;
+  if (req.headers["accept-language"]) {
+    preferredLocale = parser.pick(i18n.locales, req.headers["accept-language"]) as Maybe<string>;
+  }
```
Comment (zomars) [inline lib/core/i18n/i18n.utils.ts:15]: Type fixes

File: lib/errors.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+  if (cause instanceof Error) {
+    return cause;
+  }
+  if (typeof cause === "string") {
+    // @ts-expect-error https://github.com/tc39/proposal-error-cause
+    return new Error(cause, { cause });
+  }
+
+  return new Error(`Unhandled error of type '${typeof cause}''`);
+}
```
Comment (zomars) [inline lib/errors.ts:11]: Extracted since it's being reused in multiple files

File: package.json [hunk 8]
```diff
@@ -110,6 +110,7 @@
     "@types/react": "^17.0.18",
     "@types/react-phone-number-input": "^3.0.13",
     "@types/react-select": "^4.0.17",
+    "@types/stripe": "^8.0.417",
```
Comment (zomars) [inline package.json:113]: Added types and moved to dev dependencies

File: pages/_error.tsx [hunk 9]
```diff
@@ -25,23 +26,6 @@ type AugmentedNextPageContext = Omit<NextPageContext, "err"> & {
...
-    return cause;
-  }
-  if (typeof cause === "string") {
-    // @ts-expect-error https://github.com/tc39/proposal-error-cause
-    return new Error(cause, { cause });
-  }
-
-  return new Error(`Unhandled error of type '${typeof cause}''`);
-}
-
```
Comment (zomars) [inline pages/_error.tsx:44]: As explained above

File: pages/event-types/index.tsx [hunk 10]
```diff
@@ -223,7 +225,7 @@ const EventTypeList = ({ readOnly, types, profile }: EventTypeListProps): JSX.El
 };
 
 interface EventTypeListHeadingProps {
-  profile: Profile;
+  profile: EventTypeGroupProfile;
```
Comment (zomars) [inline pages/event-types/index.tsx:228]: Type fixing

File: pages/getting-started.tsx [hunk 11]
```diff
@@ -22,33 +15,29 @@ import TimezoneSelect from "react-timezone-select";
...
 
-type OnboardingProps = {
-  user: User;
-  integrations?: Record<string, string>[];
-  eventTypes?: EventType[];
-  schedules?: Schedule[];
-};
-
-export default function Onboarding(props: OnboardingProps) {
+export default function Onboarding(props: inferSSRProps<typeof getServerSideProps>) {
```
Comment (zomars) [inline pages/getting-started.tsx:40]: Infer prop types

File: pages/getting-started.tsx [hunk 12]
```diff
@@ -22,33 +15,29 @@ import TimezoneSelect from "react-timezone-select";
...
+import { Alert } from "@components/ui/Alert";
 import Button from "@components/ui/Button";
 import SchedulerForm, { SCHEDULE_FORM_ID } from "@components/ui/Schedule/Schedule";
 import Text from "@components/ui/Text";
-import ErrorAlert from "@components/ui/alerts/Error";
 
-import { AddCalDavIntegrationRequest } from "../lib/integrations/CalDav/components/AddCalDavIntegration";
+import getCalendarCredentials from "@server/lib/getCalendarCredentials";
+import getConnectedCalendars from "@server/lib/getConnectedCalendars";
+
```
Comment (zomars) [inline pages/getting-started.tsx:34]: Reusing the logic from integrations page

File: pages/getting-started.tsx [hunk 13]
```diff
@@ -433,11 +296,18 @@ export default function Onboarding(props: OnboardingProps) {
...
+            <>
+              <ConnectedCalendarsList connectedCalendars={props.connectedCalendars} />
+              <ShellSubHeading
+                className="mt-6"
+                title={<SubHeadingTitleWithConnections title="Connect an additional calendar" />}
+              />
+            </>
+          )}
+          <CalendarsList calendars={props.integrations} />
+        </>
```
Comment (zomars) [inline pages/getting-started.tsx:310]: Same components as integrations page

File: pages/getting-started.tsx [hunk 14]
```diff
@@ -528,11 +398,11 @@ export default function Onboarding(props: OnboardingProps) {
           setSubmitting(true);
           console.log("updating");
           await updateUser({
-            description: bioRef.current.value,
+            bio: bioRef.current?.value,
```
Comment (zomars) [inline pages/getting-started.tsx:401]: Tested and working

File: pages/getting-started.tsx [hunk 15]
```diff
@@ -572,14 +442,14 @@ export default function Onboarding(props: OnboardingProps) {
...
               </Text>
             </header>
-            <section className="space-y-2 pt-4">
+            <section className="pt-4 space-y-2">
               <Text variant="footnote">
                 Step {currentStep + 1} of {steps.length}
               </Text>
 
-              {error && <ErrorAlert {...error} />}
+              {error && <Alert severity="error" {...error} />}
```
Comment (zomars) [inline pages/getting-started.tsx:450]: Replaced deprecated component

File: pages/getting-started.tsx [hunk 16]
```diff
@@ -660,6 +530,12 @@ export async function getServerSideProps(context: NextPageContext) {
       avatar: true,
       timeZone: true,
       completedOnboarding: true,
+      selectedCalendars: {
+        select: {
+          externalId: true,
+          integration: true,
+        },
+      },
```
Comment (zomars) [inline pages/getting-started.tsx:538]: Needed to display connected calendars

File: pages/getting-started.tsx [hunk 17]
```diff
@@ -686,7 +562,14 @@ export async function getServerSideProps(context: NextPageContext) {
     },
   });
 
-  integrations = getIntegrations(credentials).map((item) => omit(item, "key"));
+  integrations = getIntegrations(credentials)
+    .filter((item) => item.type.endsWith("_calendar"))
+    .map((item) => omit(item, "key"));
```
Comment (zomars) [inline pages/getting-started.tsx:567]: Show only calendar integrations

File: pages/getting-started.tsx [hunk 18]
```diff
@@ -686,7 +562,14 @@ export async function getServerSideProps(context: NextPageContext) {
...
+
+  // get user's credentials + their connected integrations
+  const calendarCredentials = getCalendarCredentials(credentials, user.id);
+  // get all the connected integrations' calendars (from third party)
+  connectedCalendars = await getConnectedCalendars(calendarCredentials, user.selectedCalendars);
```
Comment (zomars) [inline pages/getting-started.tsx:572]: Same logic as integrations page

File: pages/reschedule/[uid].tsx [hunk 19]
```diff
@@ -44,7 +44,11 @@ export async function getServerSideProps(context: GetServerSidePropsContext) {
   const eventType = booking.eventType;
 
   const eventPage =
-    (eventType.team ? "team/" + eventType.team.slug : booking.user.username) + "/" + booking.eventType.slug;
+    (eventType.team
+      ? "team/" + eventType.team.slug
+      : booking.user?.username || "rick") /* This shouldn't happen */ +
+    "/" +
+    booking.eventType.slug;
```
Comment (zomars) [inline pages/reschedule/[uid].tsx:51]: Fixing types in edge cases

File: server/routers/viewer.tsx [hunk 20]
```diff
@@ -314,57 +313,10 @@ const loggedInViewerRouter = createProtectedRouter()
...
-            return {
-              integration,
-              credentialId,
-              error: {
-                message: error.message,
-              },
-            };
-          }
-        })
-      );
```
Comment (zomars) [inline server/routers/viewer.tsx:367]: Extracted logic to reuse

File: server/ssg.ts [hunk 21]
```diff
@@ -10,5 +10,12 @@ export const ssg = createSSGHelpers({
     prisma,
     session: null,
     user: null,
+    i18n: {
+      _nextI18Next: {
+        initialI18nStore: null,
+        userConfig: null,
+      },
+    },
+    locale: "en",
```
Comment (zomars) [inline server/ssg.ts:19]: Type fixes

File: components/integrations/ConnectIntegrations.tsx [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+
+import { AddAppleIntegrationModal } from "@lib/integrations/Apple/components/AddAppleIntegration";
+import { AddCalDavIntegrationModal } from "@lib/integrations/CalDav/components/AddCalDavIntegration";
+import { trpc } from "@lib/trpc";
+
+import { ButtonBaseProps } from "@components/ui/Button";
+
+export default function ConnectIntegration(props: {
+  type: string;
+  render: (renderProps: ButtonBaseProps) => JSX.Element;
```
Comment (KATT) [inline components/integrations/ConnectIntegrations.tsx:13]:
```suggestion
  render: (renderProps: ButtonBaseProps) => JSX.Element;
  onOpenChange: (isOpen: boolean) => void | Promise<void>;
```

File: components/integrations/ConnectIntegrations.tsx [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+  const setIsModalOpen: typeof _setIsModalOpen = (v) => {
+    _setIsModalOpen(v);
+    // refetch intergrations on modal toggles
+
+    // FIXME Find a better way to refresh data on onboarding (or migrate to tRPC)
+    if (router.pathname === "/getting-started") {
+      refreshData();
+    } else {
+      utils.invalidateQueries(["viewer.integrations"]);
+    }
```
Comment (KATT) [inline components/integrations/ConnectIntegrations.tsx:42]:
```suggestion
    _setIsModalOpen(v);
    props.onOpenChange(v)
```

File: components/integrations/ConnectedCalendarsList.tsx [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+  connectedCalendars: {
+    credentialId: number;
+    calendars?: {
+      externalId: string;
+      name: string;
+      isSelected: boolean;
+    }[];
+    error?: {
+      message: string;
+    };
```
Comment (KATT) [inline components/integrations/ConnectedCalendarsList.tsx:21]: These are mutually exclusive and can be typed something like

```tsx
type CalIntersection = {
|  {
    calendars: {
      externalId: string;
      name: string;
      isSelected: boolean;
    }[];
   error?: never;
  }
| {
    calendars?: never;
    error: {
      message: string;
    };
}
}
```

File: components/integrations/DisconnectIntegration.tsx [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+
+import { Dialog } from "@components/Dialog";
+import ConfirmationDialogContent from "@components/dialog/ConfirmationDialogContent";
+import { ButtonBaseProps } from "@components/ui/Button";
+
+export default function DisconnectIntegration(props: {
+  /** Integration credential id */
+  id: number;
+  render: (renderProps: ButtonBaseProps) => JSX.Element;
+}) {
```
Comment (KATT) [inline components/integrations/DisconnectIntegration.tsx:15]: see comment on `ConnectedCalendarsList`

File: components/ui/WeekdaySelect.tsx [hunk 26]
```diff
@@ -11,10 +16,10 @@ export const WeekdaySelect = (props) => {
     props.onSelect(activeDays.map((v, idx) => (v ? idx : -1)).filter((v) => v !== -1));
   }, [activeDays]);
 
-  const toggleDay = (e, idx: number) => {
+  const toggleDay = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>, idx: number) => {
```
Comment (KATT) [inline components/ui/WeekdaySelect.tsx:19]: If this is written inline you don't have to type this

Comment (KATT) [reply components/ui/WeekdaySelect.tsx:19]: Easier would be to do this 

```suggestion
  const toggleDay = (idx: number) => {
```

and then do the `e.preventDefault()` inline

Comment (KATT) [reply components/ui/WeekdaySelect.tsx:19]: ... or to refactor the whole thing to react-hook-form, but I understand you don't wanna do that

File: components/ui/form/radio-area/RadioAreaGroup.tsx (hunk 4 above)
Comment (KATT) [reply components/ui/form/radio-area/RadioAreaGroup.tsx:7]: Yeah, I think it needs 

- deprecating
- refactoring a copy to be more compliant to normal `<input`-fields
- moved into where the `<Input/>`

 (not as part of this PR)

File: components/integrations/CalendarsList.tsx [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+import IntegrationListItem from "./IntegrationListItem";
+
+interface Props {
+  calendars: {
+    children?: ReactNode;
+    description: string;
+    imageSrc: string;
+    title: string;
+    type: string;
+  }[];
```
Comment (KATT) [inline components/integrations/CalendarsList.tsx:16]: Can add a notifier here to extract the bespoke handling further down

```suggestion
  }[];
  onChanged: () => Promise<void> | void
```

File: pages/integrations/index.tsx [hunk 28]
```diff
@@ -672,79 +454,14 @@ export default function IntegrationsPage() {
...
-                          <Button color="secondary" {...btnProps}>
-                            Connect
-                          </Button>
-                        )}
-                      />
-                    }
-                  />
-                ))}
-              </List>
+              <CalendarsList calendars={data.calendar.items} />
```
Comment (KATT) [inline pages/integrations/index.tsx:464]:
```suggestion
              <CalendarsList calendars={data.calendar.items} onChanged={() => trpc.invalidateQueries(['integrations']} />
```

File: server/lib/getConnectedCalendars.ts [hunk 29]
```diff
@@ -0,0 +1,50 @@
...
+          error: {
+            message: error.message,
+          },
+        };
+      }
+    })
+  );
+
+  return connectedCalendars;
+}
```
Comment (KATT) [inline server/lib/getConnectedCalendars.ts:50]: Move this file into something like `/server/integrations/getConnectedCalendars.ts` as it's server-only

File: server/ssg.ts (hunk 21 above)
Comment (KATT) [reply server/ssg.ts:19]: Nice, I had to do that on another PR on our private fork too

**Review by KATT — CHANGES_REQUESTED** [review]
Very good PR - I really love that you added comments to walk reviewers through the changes, they really, really help!

The main thing I have is to decouple the handling further so components far down the tree don't explicitly call `invalidateQueries()` etc, and that you instead use a delegate pattern to notify the parent of changes for it to do the refreshing

File: components/integrations/DisconnectIntegration.tsx [hunk 30]
```diff
@@ -0,0 +1,60 @@
...
+          "Content-Type": "application/json",
+        },
+      });
+      if (!res.ok) {
+        throw new Error("Something went wrong");
+      }
+    },
+    {
+      async onSettled() {
+        props.onOpenChange(modalOpen);
```
Comment (KATT) [inline components/integrations/DisconnectIntegration.tsx:30]: makes sure the mutation doesn't leave loading until the promise is resolved
```suggestion
        await props.onOpenChange(modalOpen);
```

**Review by KATT — APPROVED** [review]
throw my suggestion in and ship it

File: components/integrations/DisconnectIntegration.tsx (hunk 30 above)
Comment (zomars) [reply components/integrations/DisconnectIntegration.tsx:30]: Sorry @KATT automerge was enabled. Will make a quick PR

**Pull request merged**